libphal/phal_pdbg.C

```diff
@@ -323,4 +326,121 @@ bool hasControlTransitionedToHost()
 	return (l_coreScratchRegData == 0);
 }
 
+std::optional<LocationCode> getUnexpandedLocCode(const LocationCode &locCode)
+{
+	// Location code should start with "U"
+	if (locCode[0] != 'U') {
+		log(level::ERROR,
+		    "Location code should start with \"U\""
+		    " but, given location code %s",
+		    locCode);
+		return std::nullopt;
+	}
+
+	// Given location code length should be need to match with minimum
+	// length
+	// to drop expanded format in given location code.
+	if (locCode.length() < EXP_LOCATIN_CODE_MIN_LENGTH) {
+		log(level::ERROR,
+		    "Given location code %s is not meet "
+		    "with minimum length %d",
+		    locCode, EXP_LOCATIN_CODE_MIN_LENGTH);
+		return std::nullopt;
+	}
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATIN_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode);
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATIN_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+int getAttrType(const LocationCode &locationCode, ATTR_TYPE_Type &fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	// Defining a list of existing frus
+	// Note: We need to update the list if a new fru is added
+	std::set<const char *> frus{"bmc", "proc", "tpm", "dimm"};
+	struct pdbg_target *target;
+
+	for (auto fru : frus) {
+		pdbg_for_each_class_target(fru, target)
+		{
+			if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode)) {
+				continue;
+			} else {
+				if (locCode == locationCode) {
+					if (DT_GET_PROP(ATTR_TYPE, target,
+							fruType)) {
+						return openpower::phal::
+						    exception::
+							DEVTREE_ATTR_READ_FAIL;
+					} else {
+						return 0; // success
+					}
+				}
+			}
+		}
+	}
+	return -1;
+}
+
+std::expected<FRUType, openpower::phal::exception::ERR_TYPE>
```

Comment:
std::expected<FRUType, openpower::phal::exception::ERR_TYPE>

we need not complicate the caller with all these stuff unless the caller is taking different action based on error code, as this is small api just return 0 or -1

Reply:
This API is defined as `noexcept`, so we should not throw any exceptions.

Why did we not prefer exceptions? The error log component aims for a best effort approach while creating a PEL. If any errors occur internally or with dependent components, they need to be captured in the same PEL debug data section to identify and fix these issues in the field or lab. Therefore, I preferred to get the error code details and messages, which can be added to the PEL. Writing try and catch blocks was not preferred because the existing PHAL error infrastructure supports fitting within std::expected, which led to making this API noexcept.

If you prefer handling this with the existing exception design for other APIs, that's fine. However, I would not prefer altering this API at the moment since its signature is frozen and has been discussed with the PHAL team. Alternatively, you could provide us with an API that is noexcept, which can be used as is, and have another API with exceptions that throws from that API. The exceptions can be caught in this API and return error codes, as is standard with having both exception and non-exception variants.